assets/src/edit-story/app/media/utils/useMediaUploadQueue.js

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useEffect, useMemo } from 'react';
+import { v4 as uuidv4 } from 'uuid';
+import {
+  trackError,
+  trackEvent,
+  getTimeTracker,
+} from '@web-stories-wp/tracking';
+
+/**
+ * Internal dependencies
+ */
+import { useUploader } from '../../uploader';
+import useReduction from '../../../utils/useReduction';
+import { revokeBlob } from '../../../utils/blobs';
+import useTranscodeVideo from './useTranscodeVideo';
+import getResourceFromAttachment from './getResourceFromAttachment';
+import getResourceFromLocalFile from './getResourceFromLocalFile';
+
+const initialState = {
+  queue: [],
+};
+
+const reducer = {
+  addItem: (
+    state,
+    {
+      payload: {
+        file,
+        resource,
+        onUploadStart,
+        onUploadProgress,
+        onUploadError,
+        onUploadSuccess,
+      },
+    }
+  ) => {
+    const id = uuidv4();
+    const newItem = {
```

Comment:
Not for now, but would we use this to have a progress bar? So a bit of state for percentage complete?

Reply:
Yeah maybe. Progress could be tracked here, but that would result in lots and lots of state updates every time progress is updated (which could easily be dozens of times). Using callbacks might be more performant.